sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BeamRowToStorageApiProto.java

```diff
@@ -0,0 +1,339 @@
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableMap;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Iterables;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Lists;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.primitives.Bytes;
+import org.joda.time.Instant;
+import org.joda.time.ReadableInstant;
+
+/**
+ * Utility methods for converting Beam {@link Row} objects to dynamic protocol message, for use with
+ * the Storage write API.
+ */
+public class BeamRowToStorageApiProto {
+  // Number of digits after the decimal point supported by the NUMERIC data type.
+  private static final int NUMERIC_SCALE = 9;
+  // Maximum and minimum allowed values for the NUMERIC data type.
+  private static final BigDecimal MAX_NUMERIC_VALUE =
+      new BigDecimal("99999999999999999999999999999.999999999");
+  private static final BigDecimal MIN_NUMERIC_VALUE =
+      new BigDecimal("-99999999999999999999999999999.999999999");
+
+  // TODO(reuvenlax): Support BIGNUMERIC and GEOGRAPHY types.
+  static final Map<TypeName, Type> PRIMITIVE_TYPES =
+      ImmutableMap.<TypeName, Type>builder()
+          .put(TypeName.INT16, Type.TYPE_INT32)
+          .put(TypeName.BYTE, Type.TYPE_INT32)
+          .put(TypeName.INT32, Type.TYPE_INT32)
+          .put(TypeName.INT64, Type.TYPE_INT64)
+          .put(TypeName.FLOAT, Type.TYPE_FLOAT)
+          .put(TypeName.DOUBLE, Type.TYPE_DOUBLE)
+          .put(TypeName.STRING, Type.TYPE_STRING)
+          .put(TypeName.BOOLEAN, Type.TYPE_BOOL)
+          .put(TypeName.DATETIME, Type.TYPE_INT64)
+          .put(TypeName.BYTES, Type.TYPE_BYTES)
+          .put(TypeName.DECIMAL, Type.TYPE_BYTES)
+          .build();
+
+  // A map of supported logical types to the protobuf field type.
+  static final Map<String, Type> LOGICAL_TYPES =
+      ImmutableMap.<String, Type>builder()
+          .put(SqlTypes.DATE.getIdentifier(), Type.TYPE_INT32)
+          .put(SqlTypes.TIME.getIdentifier(), Type.TYPE_INT64)
+          .put(SqlTypes.DATETIME.getIdentifier(), Type.TYPE_INT64)
+          .put(SqlTypes.TIMESTAMP.getIdentifier(), Type.TYPE_INT64)
+          .put(EnumerationType.IDENTIFIER, Type.TYPE_STRING)
+          .build();
+
+  static final Map<TypeName, Function<Object, Object>> PRIMITIVE_ENCODERS =
+      ImmutableMap.<TypeName, Function<Object, Object>>builder()
+          .put(TypeName.INT16, o -> Integer.valueOf((Short) o))
+          .put(TypeName.BYTE, o -> Integer.valueOf((Byte) o))
+          .put(TypeName.INT32, Functions.identity())
+          .put(TypeName.INT64, Functions.identity())
+          .put(TypeName.FLOAT, Function.identity())
+          .put(TypeName.DOUBLE, Function.identity())
+          .put(TypeName.STRING, Function.identity())
+          .put(TypeName.BOOLEAN, Function.identity())
+          .put(TypeName.DATETIME, o -> ((ReadableInstant) o).getMillis() * 1000)
```

Comment:
Datetime needs encoding, it shouldn't be a direct micros.

Reply:
DATETIME in Beam is improperly named - it's actually a timestamp, not a DATETIME

Reply:
So it is both a logical type and a primitive type?

Reply:
nvm, just realize it is DATETIME from different types.